junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryTests.java

```diff
@@ -884,4 +891,96 @@ void nested() {
 			}
 		}
 	}
+
+	static class MultipleTempDirsTestCase {
+
+		@TempDir("a")
+		static Path A;
+		static Path B;
+
+		@BeforeAll
+		static void beforeAll(@TempDir("a") Path dir1, @TempDir("b") Path dir2) {
```

Comment:
The test uses the same ID for multiple `@TempDir` paths, e.g. field `A` and parameter `dir`.
I thought that this will not be allowed in the future according to https://github.com/junit-team/junit5/issues/1967#issuecomment-850821121?

You are even mixing `static` with non-`static` scopes here, e.g. `static` `A` with parameter `dir4` of `beforeEach`.

Reply:
You're right in pointing out that I will have to change that when tackling the other issue.

Reply:
Thanks for clarification. If you don't mind I would repurpose this thread here for discussion about 1967 and this implementation (to avoid spamming all the watchers of that issue).

In your comment on 1967 you wrote:
> forbid injecting the same temp dir twice on the same level.

What is the "same level"? I can currently only think of applying this to the "same scope", including sub-scopes, i.e.:
- Multiple `static` fields must not inject the same temp dir
- `static` field temp dir cannot be injected again in instance field
- Multiple instance fields must not inject the same temp dir
- `static` or instance field temp dir cannot be injected again as parameter
- Multiple parameters of the same callable must not inject the same temp dir
- Parameters of any setup / teardown methods and any test methods must not inject the same temp dir

With these restrictions in place the definition of how `@TempDir` works can then also greatly be simplified (allowing to avoid any IDs at all): "Every element annotated with `@TempDir` creates a separate new temp directory"
This would remove the need to "guess the semantics" (as you expressed in a comment on 1967).

Your ID approach would have one minor advantage though (assuming you indeed disallow multiple `@TempDir`s to inject the same directory): Existing unit tests which use multiple `@TempDir` to refer to the same directory would break cleanly with a concise exception message. With my proposed change above (removal of ID), such unit tests might fail with confusing assertion exceptions (since the tests themselves, instead of JUnit test setup, would fail). However, then again, how likely is it that such unit test code exists? Maybe we are trying to consider a case here which in reality never or extremely rarely exists, at the cost of making the revised `@TempDir` for all future users more complicated.

Additionally your comment "That's exactly how the new solution works" on 1967 is confusing me. I though that this will be forbidden since injecting the same temp directory will not be possible.
And really, that user should (probably) declare a `static` `@TempDir` field instead of reusing the same ID. With a field you get IDE support and can easily see which test methods access this field. When using an ID for this you have no direct IDE support at all, you have to perform plain text searches in your code, which is rather awful and error prone in case formatting of annotations is not consistent (e.g. different amount of whitespaces).

I have yet to see a use case where multiple `@TempDir` refering to the same directory gives any advantage compared to using an instance or `static` field instead. (Though maybe there exist some, feel free to point me to any of them.)

Reply:
We shouldn't be discussing this here but on #2219. 😉 

> Parameters of any setup / teardown methods and any test methods must not inject the same temp dir

I agree with your list of restrictions except for this one. I don't think we should break the following valid scenario:

```java
class ValidTest {

	@BeforeAll
	static void beforeAll(@TempDir Path tempDir) {
		// prepare files all tests need but won't change
	}

	@BeforeEach
	void beforeEach(@TempDir Path tempDir) {
		// prepare files that may be destroyed by each test
	}

	@Test
	void testRequiringTempDir(@TempDir Path tempDir) {
		// use tempDir
	}

	@Test
	void anotherTestRequiringTempDir(@TempDir Path tempDir) {
		// use tempDir
	}

	@Test
	void oneOffTestNotRequiringTempDir() {
		System.out.println("b = " + b);
		System.out.println("c = " + c);
	}
}
```

> Additionally your comment "That's exactly how the new solution works" on 1967 is confusing me. I though that this will be forbidden since injecting the same temp directory will not be possible.

see above example

Reply:
Personally I think for this specific example it would be cleaner to have one single `tempDir` `static ` field and then have every method access that field instead. This would make the scope of the temp directory clearer and it would be obvious which methods share this directory. The fact that a parameter is somehow linked to a parameter of another method (e.g. `beforeAll`'s `tempDir` and `testRequiringTempDir`'s `tempDir`) goes against the normal scope of parameters which is per method. 

(Though maybe this boils down to personal code style preferences?)